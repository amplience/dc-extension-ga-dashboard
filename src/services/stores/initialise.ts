import {
  contentItemFilter,
  editionFilter,
  slotFilter,
} from '../../stores/ga-query-filters';
import {
  breakdown,
  contentItemIdMapping,
  editionIdMapping,
  gaApiKey,
  gaClientEmail,
  gaClientId,
  gaViewId,
  slotIdMapping,
} from '../../stores/google-analytics';
import { currencyCode, locale } from '../../stores/localization';
import type { ExtensionInstallation } from '../extension-sdk/extension-sdk.service';

export function initialiseStores(config: ExtensionInstallation): void {
  gaViewId.set(config?.googleAnalyticsViewId || '');
  gaClientId.set(config?.googleAnalyticsClientId || '');
  gaClientEmail.set(config?.googleAnalyticsClientEmail);
  gaApiKey.set(config?.googleAnalyticsKey || '');

  contentItemIdMapping.set(config?.mappings?.contentItemId);
  editionIdMapping.set(config?.mappings?.editionId);
  slotIdMapping.set(config?.mappings?.slotId);

  contentItemFilter.set(config?.filters?.contentItemFilter);
  editionFilter.set(config?.filters?.editionFilter);
  slotFilter.set(config?.filters?.slotFilter);

  breakdown.set(
    config?.breakdown || {
      dimension: 'ga:deviceCategory',
      title: 'Device Breakdown',
    }
  );

  if (config?.localization?.locale) {
    locale.set(config?.localization?.locale);
  }

  if (config?.localization?.currencyCode) {
    currencyCode.set(config?.localization?.currencyCode);
  }
}
