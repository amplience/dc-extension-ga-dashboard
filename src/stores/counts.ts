import { writable, derived } from 'svelte/store';
import type {
  SearchIndexUsersCount,
  SearchIndexSearchesCount,
  SearchIndexNoResultsRate,
} from 'dc-management-sdk-js';

export interface Counts {
  usersCount?: SearchIndexUsersCount;
  searchesCount?: SearchIndexSearchesCount;
  noResultsRate?: SearchIndexNoResultsRate;
}

export const counts = writable<Counts>({});

export const searchesPerUser = derived(counts, ($counts) => {
  return $counts?.usersCount?.count && $counts?.searchesCount?.count
    ? Math.round(
        ($counts?.searchesCount?.count / $counts?.usersCount?.count) * 1e1
      ) / 1e1
    : 0;
});
