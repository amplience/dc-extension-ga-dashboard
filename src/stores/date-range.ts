import { writable } from './persisted-store';
import { formatDateAsISOString } from '../utils/date-format';

export const DAY = 24 * 60 * 60 * 1000;

export interface DateRange {
  from: string;
  to: string;
}

const NOW = new Date().setHours(0, 0, 0, 0).valueOf();

export const dateRange = writable<DateRange>('date-range', {
  from: formatDateAsISOString(new Date(NOW - DAY * 30)),
  to: formatDateAsISOString(new Date(NOW - DAY)),
});
