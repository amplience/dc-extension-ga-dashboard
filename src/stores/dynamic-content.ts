import type { DynamicContent, Hub } from 'dc-management-sdk-js';
import { writable } from 'svelte/store';

export const client = writable<DynamicContent>(null);

export const hub = writable<Hub>(null);
