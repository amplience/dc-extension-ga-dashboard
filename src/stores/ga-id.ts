import { writable as persistedWritable } from './persisted-store';

export const gaId = persistedWritable<string>('ga-id', null);
