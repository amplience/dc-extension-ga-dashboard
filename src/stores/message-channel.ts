import { writable } from 'svelte/store';
import type { ClientConnection } from 'message-event-channel';

export const connection = writable<ClientConnection>(null);
