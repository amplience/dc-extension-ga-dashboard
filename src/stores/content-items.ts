import { get, writable } from 'svelte/store';
import { client } from './dynamic-content';

function createContentItems() {
  const { subscribe, set, update } = writable([]);

  return {
    subscribe,
    fetch: async (id) => {
      const cachedContentItem = get(contentItems).find(
        (item) => item.id === id
      );
      if (cachedContentItem) {
        return cachedContentItem;
      }
      const contentItem = await get(client).contentItems.get(id);
      update((n) => [...n, ...[contentItem]]);
      return contentItem;
    },
    reset: () => set([]),
  };
}

export const contentItems = createContentItems();
