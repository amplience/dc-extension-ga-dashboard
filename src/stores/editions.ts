import { get, writable } from 'svelte/store';
import { client } from './dynamic-content';

function createEditions() {
  const { subscribe, set, update } = writable([]);

  return {
    subscribe,
    fetch: async (id) => {
      const cachedEdition = get(editions).find((item) => item.id === id);
      if (cachedEdition) {
        return cachedEdition;
      }
      const edition = await get(client).editions.get(id);
      update((n) => [...n, ...[edition]]);
      return edition;
    },
    reset: () => set([]),
  };
}

export const editions = createEditions();
