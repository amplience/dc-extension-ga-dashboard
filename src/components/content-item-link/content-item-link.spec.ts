import { tick } from 'svelte';
import { render } from '@testing-library/svelte';
import ContentItemLink from './content-item-link.svelte';
import getContentItemProperties from './services/get-content-item-properties.service';
import { connection } from '../../stores/message-channel';
import type { ClientConnection } from 'message-event-channel';

jest.mock('./services/get-content-item-properties.service');
describe('ContentItemLink', () => {
  const contentItemId = 'abc';
  const mockConnection = {
    emit: jest.fn(),
  };
  beforeEach(() => {
    connection.set((mockConnection as unknown) as ClientConnection);
    (getContentItemProperties as jest.Mock).mockReturnValue(
      Promise.resolve(['CONTENT_ITEM_URL', 'LABEL'])
    );
  });

  it('should display the content item id', () => {
    const { container } = render(ContentItemLink, {
      contentItemId,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display a the content item link with a href and label', async () => {
    const { container } = render(ContentItemLink, {
      contentItemId,
    });
    await tick();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display without a href if it was unable to get the URL', async () => {
    (getContentItemProperties as jest.Mock).mockReturnValue(
      Promise.resolve([undefined, 'LABEL'])
    );

    const { container } = render(ContentItemLink, {
      contentItemId,
    });
    await tick();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display the content item id if it was unable to get the content item', async () => {
    (getContentItemProperties as jest.Mock).mockReturnValue(
      Promise.resolve(['CONTENT_ITEM_URL', 'abc'])
    );
    const { container } = render(ContentItemLink, {
      contentItemId,
    });
    await tick();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call the connection emit function when clicked', async () => {
    (getContentItemProperties as jest.Mock).mockReturnValue(
      Promise.resolve(['CONTENT_ITEM_URL', 'LABEL'])
    );
    const { container } = render(ContentItemLink, {
      contentItemId,
    });
    await tick();
    expect(container.firstChild).toMatchSnapshot();

    const contentItemLink = document.querySelector('a');
    (contentItemLink as HTMLElement).click();

    expect(mockConnection.emit).toHaveBeenCalledWith(
      'content-item-redirect',
      contentItemId
    );
  });
});
