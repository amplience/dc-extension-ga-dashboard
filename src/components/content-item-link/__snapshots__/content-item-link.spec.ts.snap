// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentItemLink should call the connection emit function when clicked 1`] = `
<div>
  <a
    class="svelte-1xffnc5"
    href="CONTENT_ITEM_URL"
  >
    LABEL
  </a>
  
</div>
`;

exports[`ContentItemLink should display a the content item link with a href and label 1`] = `
<div>
  <a
    class="svelte-1xffnc5"
    href="CONTENT_ITEM_URL"
  >
    LABEL
  </a>
  
</div>
`;

exports[`ContentItemLink should display the content item id 1`] = `
<div>
  
  <span>
    abc
  </span>
</div>
`;

exports[`ContentItemLink should display the content item id if it was unable to get the content item 1`] = `
<div>
  <a
    class="svelte-1xffnc5"
    href="CONTENT_ITEM_URL"
  >
    abc
  </a>
  
</div>
`;

exports[`ContentItemLink should display without a href if it was unable to get the URL 1`] = `
<div>
  <a
    class="svelte-1xffnc5"
  >
    LABEL
  </a>
  
</div>
`;
